### DISCUSSION TITLE
DatabaseChanges check connection status

### DISCUSSION THREAD

**Discussion by jmajoor** [ravendb/ravendb#13048] [Q&A]
We are using DatabaseChanges subscriptions and want to make sure that the connection is active and working.
In order to do this we call "EnsureConnectionNow()", which will give us the Task of the subscription.
If this task is in a Faulted or Cancelled state, then we dispose our subscription and create a new subscription. Fortunately this hardly ever happens as the DatabaseChanges will reconnect internally most of the time.
We do not check the Connected property as this may not reflect an actual error, just an intermediate state.

Looking at the code we noticed that there are 2 paths in the DoWork that don't inform the task about it actual state.
If there is an OperationCanceledException in the ProcessChanges method, then it simply returns out of the DoWork task and it does not set the Connected flag nor mark the task as cancelled.

If there is an other exception then it will try to re-connect, but if that reconnection fails it will simply throw an exception and not set that exception on the task (the task completion source).

We did ran into a situation where the subscription seems no longer connected, yet the task is not in a faulted or cancelled state.
My question is whether our approach for checking for connectivity is correct and / or our analysis of the DatabaseChanges code is correct (i.e. it may require some improvements).

**Comment by garayx** [comment]
Hi 

Can you attach the code that you are using for analysis of the DatabaseChanges  ?

**Comment by ayende** [comment]
What build are you using? `EnsureConnectionNow()` is from 3.5, IIRC.
There is the `ConnectionStatusChanged` you can use.

**Reply by jmajoor** [reply]
The code was originally written for 3.5, but we are currently on 5.0. The `ConnectionStatusChanged` has similar problems. There are several places where the connection status changed is not called from within the DoWork task. In addition the connection status changed is also called when the client tries to re-connect internally.

I don't want to dispose my subscription if the client is already trying to re-connect. I want to know whether or not the DoWork task has ended (ie. there is no longer any attempt to re-connect) and in that case I can then dispose my subscription and create a new subscription. The only way I see is by using the EnsureConnectionNow as that gives me access to the source completion task.

The error state of DoWork is not propagated to the OnError of my Rx observable either, so there is a chance that I never receive change events if the DatabaseChanges is disconnected and has basically stopped.

**Reply by ayende** [reply]
What is it that you are actually trying to do? 
The scenario for `Changes()` is a typically optimistic one, because if there is a reconnection attempt, you'll miss up any changes that happened in the middle.

**Reply by jmajoor** [reply]
I know that the changes API is not a guaranteed delivery system. However we ran into a situation where the changes subscription had disconnected internally and there is no notification of that and no more internal reconnection attempts  (ie.. we were never going to get any event). Therefore we periodically check the changes subscription using the ensureconnectnow task to see whether the changes subscription has internally faulted and if so we create a new subscription to make sure we can keep receiving events.

We also have a periodic query that we run to make sure we didn't miss any important changes,, but like the ability to more quickly react to some changes using the changes api.

**Comment by jmajoor** [comment]
```
while (_cts.IsCancellationRequested == false)
            {
                try
                {
                    if (Connected == false)
                    {
                        _url = new Uri($"{_serverNode.Url}/databases/{_database}/changes"
                            .ToLower()
                            .ToWebSocketPath(), UriKind.Absolute);

                        await _client.ConnectAsync(_url, _cts.Token).ConfigureAwait(false);
                        wasConnected = true;
                        Interlocked.Exchange(ref _immediateConnection, 1);

                        foreach (var counter in _counters)
                        {
                            counter.Value.Set(counter.Value.OnConnect());
                        }

                        ConnectionStatusChanged?.Invoke(this, EventArgs.Empty);
                    }

                    await ProcessChanges().ConfigureAwait(false);
                }
                catch (OperationCanceledException e)
                {
                    NotifyAboutError(e);
                    // Missing?: _tcs.TrySetCanceled();
                    return;
                }
                catch (ChangeProcessingException)
                {
                    continue;
                }
                catch (Exception e)
                {
                    //We don't report this error since we can automatically recover from it and we can't
                    // recover from the OnError accessing the faulty WebSocket.
                    try
                    {
                        if (wasConnected)
                            ConnectionStatusChanged?.Invoke(this, EventArgs.Empty);

                        wasConnected = false;
                        try
                        {
                            _serverNode = await _requestExecutor.HandleServerNotResponsive(_url.AbsoluteUri, _serverNode, _nodeIndex, e).ConfigureAwait(false);
                        }
                        catch (DatabaseDoesNotExistException databaseDoesNotExistException)
                        {
                            e = databaseDoesNotExistException;
                            throw;
                        }
                        catch (Exception)
                        {
                            //We don't want to stop observe for changes if server down. we will wait for one to be up
                        }

                        if (ReconnectClient() == false)
                            return;
                    }
                    catch
                    {
                        // we couldn't reconnect
                        NotifyAboutError(e);
                    // Missing?: _tcs.TrySetException();
                        throw;
                    }
                }
```
The `OperationCanceledException` doesn't call `_tcs.TrySetCanceled();`
Also in the `catch` there are places where it returns or re-throws an exception without calling ` _tcs.TrySetException(e);`.
I marked up the code with `Missing?:`, where I think the `_tcs` needs to be updated.

**Reply by ayende** [reply]
Would you be able to send a PR for this?

**Reply by jmajoor** [reply]
PR https://github.com/ravendb/ravendb/pull/13066

However since there is no YouTrack number for this it fails with.
- commit/message/conventions - all commit messages (except in merge commits) must contain an issue number from our YouTrack e.g. 'RavenDB-1234 Fixed issue with something'
